### PR TITLE
Closes #1163 Running Tests on OSX/Windows

### LINKED ISSUE #1163: Developing on Mac (and potentially windows)
In order to reduce the friction to developing on POCS on OSX, I have figured out how to install the dependencies for POCS on a mac in a reproducible and isolated way and how to modify the test setup to work.

**Describe the solution you'd like**
Setup Steps on a Mac
```
# Create a virtualenv
python3 -m venv <path/to/venv>
source <path/to/venv>/bin/activate
# wherever the root of the repo is on your machine
cd $POCS_HOME 
pip install .[google,focuser,testing,sensors]
```
Combined with a change to `conftest.py` to address multiprocessing using spawn on windows and osx instead of fork by default
Should allow you to run pytest as usual.

**Describe alternatives you've considered**
Since we are moving away from docker, this is the 2nd best option in my opinion to make it really easy for contributors to be able to do the common pattern with open source projects. Which is to clone a repo, make a change, run tests to validate the change is working as expected, create a PR. Without having to jump through hoops or have intimate knowledge of how the project works.

### PR DESCRIPTION
**Pull request by programatt**
Update documentation for instructions on how to run tests locally on OSX. Fix conftest.py to not error when running pytest on OSX and Windows.

## Description
When attempting to run pytest on my macbook, I ran into an issue on startup where the config_server throws a strange error. 

```
W 01-20 03:15:11.281 Problem with get_config: ConnectionError(MaxRetryError("HTTPConnectionPool(host='localhost', port=6563): Max retries exceeded with url: /get-config (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x160521490>: Failed to establish a new connection: [Errno 61] Connection refused'))"))
I 01-20 03:15:11.285 Setting up the config server.
I 01-20 03:15:11.285 Starting panoptes-config-server with  config_file='tests/testing.yaml'
S 01-20 03:15:11.337 Config server Loaded 15 top-level items
S 01-20 03:15:11.337 {'name': 'Testing PANOPTES Unit', 'pan_id': 'PAN000', 'pocs': ordereddict([('INITIALIZED', False), ('CONNECTED', False), ('INTERRUPTED', False)]), 'location': ordereddict([('name', 'Mauna Loa Observatory'), ('latitude', <Quantity 19.54 deg>), ('longitude', <Quantity -155.58 deg>), ('elevation', <Quantity 3400. m>), ('horizon', <Quantity 30. deg>), ('flat_horizon', <Quantity -6. deg>), ('focus_horizon', <Quantity -12. deg>), ('observe_horizon', <Quantity -18. deg>), ('obstructions', []), ('timezone', 'US/Hawaii'), ('gmt_offset', -600)]), 'directories': ordereddict([('base', '/Users/matt/code/POCS'), ('images', '/Users/matt/code/POCS/images'), ('data', '/Users/matt/code/POCS/data'), ('resources', '/Users/matt/code/POCS/resources'), ('fields', '/Users/matt/code/POCS/conf_files/fields'), ('mounts', '/Users/matt/code/POCS/resources/mounts')]), 'db': ordereddict([('name', 'panoptes_testing'), ('type', 'file')]), 'scheduler': ordereddict([('type', 'panoptes.pocs.scheduler.dispatch'), ('fields_file', 'simulator.yaml'), ('check_file', False), ('iers_url', 'https://storage.googleapis.com/panoptes-resources/iers/ser7.dat'), ('constraints', [ordereddict([('name', 'panoptes.pocs.scheduler.constraint.Altitude')]), ordereddict([('name', 'panoptes.pocs.scheduler.constraint.MoonAvoidance'), ('options', ordereddict([('separation', 15)]))]), ordereddict([('name', 'panoptes.pocs.scheduler.constraint.Duration')])])]), 'mount': ordereddict([('brand', 'ioptron'), ('model', 'cem40'), ('driver', 'panoptes.pocs.mount.simulator'), ('serial', ordereddict([('port', '/dev/ttyUSB0'), ('timeout', 0.0), ('baudrate', 9600)])), ('non_sidereal_available', True)]), 'pointing': ordereddict([('auto_correct', True), ('threshold', 500), ('exptime', 30), ('max_iterations', 1)]), 'cameras': ordereddict([('defaults', ordereddict([('primary', 'None'), ('auto_detect', False), ('file_extension', 'fits'), ('compress_fits', True), ('make_pretty_images', True), ('keep_jpgs', False), ('readout_time', 0.5), ('timeout', 10), ('filter_type', 'RGGB'), ('cooling', ordereddict([('enabled', True), ('temperature', ordereddict([('target', 0), ('tolerance', 0.1)])), ('stable_time', 60), ('check_interval', 5), ('timeout', 300)])), ('focuser', ordereddict([('enabled', True), ('autofocus_seconds', 0.1), ('autofocus_size', 500), ('autofocus_keep_files', False)]))])), ('devices', [ordereddict([('model', 'panoptes.pocs.camera.simulator.dslr.Camera'), ('name', 'dslr.00'), ('port', '/dev/fake/dslr.00'), ('readout_time', 0.5)]), ordereddict([('model', 'panoptes.pocs.camera.simulator.dslr.Camera'), ('name', 'dslr.focuser.cooling.00'), ('port', '/dev/fake/dslr.focuser.cooling.00'), ('cooling', ordereddict([('enabled', True), ('target', 1), ('tolerance', 0.1), ('stable_time', 1), ('check_interval', 0.5), ('timeout', 2.5)])), ('focuser', ordereddict([('model', 'panoptes.pocs.focuser.simulator.Focuser'), ('focus_port', '/dev/fake/focuser.00'), ('initial_position', 20000), ('autofocus_range', [40, 80]), ('autofocus_step', [10, 20]), ('autofocus_seconds', 0.1), ('autofocus_size', 500), ('autofocus_keep_files', False)]))]), ordereddict([('model', 'panoptes.pocs.camera.simulator.dslr.Camera'), ('name', 'dslr.filterwheel.cooling.00'), ('port', '/dev/fake/dslr.filterwheel.cooling.00'), ('cooling', ordereddict([('enabled', True), ('target', 0), ('tolerance', 0.1), ('stable_time', 60), ('check_interval', 5), ('timeout', 300)])), ('filterwheel', ordereddict([('model', 'panoptes.pocs.filterwheel.simulator.FilterWheel'), ('filter_names', ['one', 'deux', 'drei', 'quattro', 'blank']), ('move_time', 0.1), ('timeout', 0.5)]))]), ordereddict([('model', 'panoptes.pocs.camera.simulator.ccd.Camera'), ('name', 'ccd.filterwheel.focuser.cooling.00'), ('serial_number', 'ccd.filterwheel.focuser.cooling.00'), ('cooling', ordereddict([('enabled', True), ('target', 0), ('tolerance', 0.1), ('stable_time', 60), ('check_interval', 5), ('timeout', 300)])), ('focuser', ordereddict([('model', 'panoptes.pocs.focuser.simulator.Focuser'), ('focus_port', '/dev/fake/focuser.00'), ('initial_position', 20000), ('autofocus_range', [40, 80]), ('autofocus_step', [10, 20]), ('autofocus_seconds', 0.1), ('autofocus_size', 500), ('autofocus_keep_files', False)])), ('filterwheel', ordereddict([('model', 'panoptes.pocs.filterwheel.simulator.FilterWheel'), ('filter_names', ['one', 'deux', 'drei', 'quattro', 'blank']), ('move_time', 0.1), ('timeout', 0.5), ('dark_position', 'blank'), ('focus_offsets', ordereddict([('one', 0), ('deux', 1), ('drei', 2), ('quattro', 3)]))]))])])]), 'observations': ordereddict([('make_timelapse', True), ('record_observations', True)]), 'panoptes_network': ordereddict([('image_storage', True), ('service_account_key', None), ('project_id', 'panoptes-survey'), ('buckets', ordereddict([('images', 'panoptes-survey')]))]), 'state_machine': 'panoptes', 'environment': ordereddict([('auto_detect', True)]), 'weather': ordereddict([('aag_cloud', ordereddict([('serial_port', '/dev/ttyUSB1')]))]), 'config_server': {'running': True}}
I 01-20 03:15:11.337 Config items saved to flask config-server
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/Users/matt/code/venvs/pocs/lib/python3.9/site-packages/_pytest/main.py", line 265, in wrap_session
INTERNALERROR>     config._do_configure()
INTERNALERROR>   File "/Users/matt/code/venvs/pocs/lib/python3.9/site-packages/_pytest/config/__init__.py", line 982, in _do_configure
INTERNALERROR>     self.hook.pytest_configure.call_historic(kwargs=dict(config=self))
INTERNALERROR>   File "/Users/matt/code/venvs/pocs/lib/python3.9/site-packages/pluggy/_hooks.py", line 277, in call_historic
INTERNALERROR>     res = self._hookexec(self.name, self.get_hookimpls(), kwargs, False)
INTERNALERROR>   File "/Users/matt/code/venvs/pocs/lib/python3.9/site-packages/pluggy/_manager.py", line 80, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
INTERNALERROR>   File "/Users/matt/code/venvs/pocs/lib/python3.9/site-packages/pluggy/_callers.py", line 60, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File "/Users/matt/code/venvs/pocs/lib/python3.9/site-packages/pluggy/_result.py", line 60, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "/Users/matt/code/venvs/pocs/lib/python3.9/site-packages/pluggy/_callers.py", line 39, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "/Users/matt/code/POCS/conftest.py", line 65, in pytest_configure
INTERNALERROR>     config_server(config_file, host=host, port=port, load_local=False, save_local=False)
INTERNALERROR>   File "/Users/matt/code/venvs/pocs/lib/python3.9/site-packages/panoptes/utils/config/server.py", line 117, in config_server
INTERNALERROR>     server_process.start()
INTERNALERROR>   File "/opt/homebrew/Cellar/python@3.9/3.9.9/Frameworks/Python.framework/Versions/3.9/lib/python3.9/multiprocessing/process.py", line 121, in start
INTERNALERROR>     self._popen = self._Popen(self)
INTERNALERROR>   File "/opt/homebrew/Cellar/python@3.9/3.9.9/Frameworks/Python.framework/Versions/3.9/lib/python3.9/multiprocessing/context.py", line 224, in _Popen
INTERNALERROR>     return _default_context.get_context().Process._Popen(process_obj)
INTERNALERROR>   File "/opt/homebrew/Cellar/python@3.9/3.9.9/Frameworks/Python.framework/Versions/3.9/lib/python3.9/multiprocessing/context.py", line 284, in _Popen
INTERNALERROR>     return Popen(process_obj)
INTERNALERROR>   File "/opt/homebrew/Cellar/python@3.9/3.9.9/Frameworks/Python.framework/Versions/3.9/lib/python3.9/multiprocessing/popen_spawn_posix.py", line 32, in __init__
INTERNALERROR>     super().__init__(process_obj)
INTERNALERROR>   File "/opt/homebrew/Cellar/python@3.9/3.9.9/Frameworks/Python.framework/Versions/3.9/lib/python3.9/multiprocessing/popen_fork.py", line 19, in __init__
INTERNALERROR>     self._launch(process_obj)
INTERNALERROR>   File "/opt/homebrew/Cellar/python@3.9/3.9.9/Frameworks/Python.framework/Versions/3.9/lib/python3.9/multiprocessing/popen_spawn_posix.py", line 47, in _launch
INTERNALERROR>     reduction.dump(process_obj, fp)
INTERNALERROR>   File "/opt/homebrew/Cellar/python@3.9/3.9.9/Frameworks/Python.framework/Versions/3.9/lib/python3.9/multiprocessing/reduction.py", line 60, in dump
INTERNALERROR>     ForkingPickler(file, protocol).dump(obj)
INTERNALERROR> AttributeError: Can't pickle local object 'config_server.<locals>.start_server'
```
Doing some searching it looks like python 3.8 on OSX changed to use spawn instead of fork for processes by default. See https://github.com/pytest-dev/pytest-flask/issues/104#issuecomment-577908228 

So I modified conftest to detect the OS and change the behavior on windows and osx and the pytest command runs as expected.

## Related Issue

#1163 

## How Has This Been Tested?
The changes allowed pytest to run successfully on a mac

## Screenshots (if appropriate):
n/a

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] My code follows the code style of this project.
- [X] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [X] I have read the **CONTRIBUTING** document.
- [X] I have added tests to cover my changes.

### CODE REVIEW

File: conftest.py [hunk 1]
```diff
@@ -3,6 +3,14 @@
 import shutil
 import stat
 import tempfile
+from sys import platform
+
+# Windows and OSX use spawn instead of fork by default
+# which causes the config server to fail to start
+if platform == "darwin" or platform == "win32":
```
Comment (programatt) [inline conftest.py:10]: I think you could make the argument that this belongs in the config server code in panoptes-utils too. @wtgee let me know if you'd rather see it over there.

File: README.md [hunk 2]
```diff
@@ -130,6 +130,10 @@ See [Coding in PANOPTES](https://github.com/panoptes/POCS/wiki/Coding-in-PANOPTE
 #### [Testing]
 
 To test the software, you can use the standard `pytest` tool from the root of the directory.
+
+To install the dependencies for testing, you can run `pip install .[google,testing,sensors,focuser]` from the root
```
Comment (wtgee) [inline README.md:134]: Many shells will have problems with the `[` without quoting so let's just put this in there.

```suggestion
To install the dependencies for testing, you can run `pip install ".[google,testing,sensors,focuser]"` from the root
```

**Review by wtgee — CHANGES_REQUESTED** [review]
Thanks @programatt, I just approved your first-time contribution so the tests are running now.

Can you add windows and mac to the `matrix.os` in `.github/workflows/pythontest.yaml` as part of this PR?

**Comment by programatt** [comment]
@wtgee anything else needed?

File: .github/workflows/pythontest.yaml [hunk 3]
```diff
@@ -3,10 +3,10 @@ name: Test Python
 on: [ push, pull_request ]
 jobs:
   lint:
-    runs-on: ubuntu-latest
+    runs-on: "ubuntu-latest"
     strategy:
       matrix:
-        python-version: [ "3.10" ]
+        python-version: [ "3.10" ]
```
Comment (wtgee) [inline .github/workflows/pythontest.yaml:9]: Editor not trimming whitespace?

File: conftest.py [hunk 4]
```diff
@@ -13,6 +21,8 @@
 from panoptes.utils.config.client import set_config
 from panoptes.utils.config.server import config_server
 
+
+
```
Comment (wtgee) [inline conftest.py:25]: Were these editor added or intentional?

```suggestion
```

**Review by wtgee — APPROVED** [review]
LGTM, although the macos tests are failing on install. Looks like they are using python 2.7 by default? Which is odd. Must be how the interpreter needs to be specified or something.

**Comment by programatt** [comment]
@wtgee looks like the mac tests are now failing with this error
```
___________ ERROR collecting src/panoptes/pocs/filterwheel/libefw.py ___________
src/panoptes/pocs/filterwheel/libefw.py:12: in <module>
    class EFWDriver(AbstractSDKDriver):
src/panoptes/pocs/filterwheel/libefw.py:15: in EFWDriver
    _libudev = load_c_library('udev', mode=ctypes.RTLD_GLOBAL)
/usr/local/lib/python3.9/site-packages/panoptes/utils/library.py:39: in load_c_library
    raise error.NotFound(f"Cound not find {name} library!")
E   panoptes.utils.error.NotFound: NotFound: Cound not find udev library!
```

Have you seen that before? I'm guessing theres a library that needs to be added to the mac gh action image.

**Comment by wtgee** [comment]
Better late than never?

Fixed this upstream at https://github.com/panoptes/panoptes-utils/pull/304

**Pull request closed without merging**